barretenberg/cpp/pil/vm2/addressing.pil

```diff
@@ -1,15 +1,295 @@
+#[INDIRECT_FROM_MEMORY_2]
+sel_should_apply_indirection[2] { /*address=*/op_after_relative[2], /*value=*/rop[2], /*tag=*/rop_tag[2], /*rw=*/precomputed.zero/*(read)*/ } in memory.sel { memory.address, memory.value, memory.tag, memory.rw };
+#[INDIRECT_FROM_MEMORY_3]
+sel_should_apply_indirection[3] { /*address=*/op_after_relative[3], /*value=*/rop[3], /*tag=*/rop_tag[3], /*rw=*/precomputed.zero/*(read)*/ } in memory.sel { memory.address, memory.value, memory.tag, memory.rw };
+#[INDIRECT_FROM_MEMORY_4]
+sel_should_apply_indirection[4] { /*address=*/op_after_relative[4], /*value=*/rop[4], /*tag=*/rop_tag[4], /*rw=*/precomputed.zero/*(read)*/ } in memory.sel { memory.address, memory.value, memory.tag, memory.rw };
+#[INDIRECT_FROM_MEMORY_5]
+sel_should_apply_indirection[5] { /*address=*/op_after_relative[5], /*value=*/rop[5], /*tag=*/rop_tag[5], /*rw=*/precomputed.zero/*(read)*/ } in memory.sel { memory.address, memory.value, memory.tag, memory.rw };
+#[INDIRECT_FROM_MEMORY_6]
+sel_should_apply_indirection[6] { /*address=*/op_after_relative[6], /*value=*/rop[6], /*tag=*/rop_tag[6], /*rw=*/precomputed.zero/*(read)*/ } in memory.sel { memory.address, memory.value, memory.tag, memory.rw };
+
+// Otherwise, we propagate the operands from the previous step.
+#[INDIRECT_PROPAGATION_0]
+(1 - sel_should_apply_indirection[0]) * (rop[0] - op_after_relative[0]) = 0;
+#[INDIRECT_PROPAGATION_1]
+(1 - sel_should_apply_indirection[1]) * (rop[1] - op_after_relative[1]) = 0;
+#[INDIRECT_PROPAGATION_2]
+(1 - sel_should_apply_indirection[2]) * (rop[2] - op_after_relative[2]) = 0;
+#[INDIRECT_PROPAGATION_3]
+(1 - sel_should_apply_indirection[3]) * (rop[3] - op_after_relative[3]) = 0;
+#[INDIRECT_PROPAGATION_4]
+(1 - sel_should_apply_indirection[4]) * (rop[4] - op_after_relative[4]) = 0;
+#[INDIRECT_PROPAGATION_5]
+(1 - sel_should_apply_indirection[5]) * (rop[5] - op_after_relative[5]) = 0;
+#[INDIRECT_PROPAGATION_6]
+(1 - sel_should_apply_indirection[6]) * (rop[6] - op_after_relative[6]) = 0;
+
+// Operands after indirect resolution are the resolved_operands rop[i], ... (these are defined in execution.pil).
+
+/**************************************************************************************************
+ *  Final Guarantees
+ **************************************************************************************************/
+
+// This error is true iff some final check failed. That is if some tag is not U32.
+// Observe that we don't need to know exactly which one failed.
+// We use this fact to "batch" the checks and do only 1 comparison against 0 (inverse check).
+pol commit sel_some_final_check_failed;
+// Each tag takes at most 3 bits (guaranteed by mem lookup!), we can encode all of them in a field.
+pol BATCHED_TAGS = sel_op_is_address[0] * sel_should_apply_indirection[0] * 2**0  * rop_tag[0]
```

Comment:
I asked Jean about this and he thinks it works, but be critical :)

Reply:
@jeanmon I think I will still modify things to
- just be 1 expression that has the subtraction inside the 2**n term (instead of having expressions, that should be cheaper)
- have the term be something like `k + rop_tag[i] - MEM_U32`. Given `rop_tag` between 0 and 6 I want to be sure that whatever term I put in there is also between 0 and 6. I get concerned if the subtraction wraps because then the term becomes huge (certainly bigger than 3 bits) and it could modify other bits. You said this doesn't matter but it feels safer